New entry swift collections (#3912)
Codecademy/docs: content/swift/concepts/collections/collections.md

## Example

The following example initializes the `colors` array with four color names. Then, it showcases how to access elements by index. Afterwards, using the `append` method adds a new element to the `colors` array. Then, the example demonstrates removing an element from the array with the `remove` method. Finally, the example shows how to iterate through the `colors` array using a `for-in` loop and prints each color individually.

```swift
// Define an array of strings
var colors: [String] = ["Yellow", "Purple", "Orange", "Green"]

// Access and print elements
print(colors[0])  // Output: Yellow
print(colors[2])  // Output: Orange

// Add an element
colors.append("Red")
print(colors)  // Output: ["Yellow", "Purple", "Orange", "Green", "Red"]

// Remove an element
colors.remove(at: 1)
print(colors)  // Output: ["Yellow", "Orange", "Green", "Red"]

// Iterate through the array
for color in colors {
    print(color)
}

// Output:
// Yellow
// Orange
// Green

> // Red

Author: I created a new output block for all the outputs.

## Performance of Collections

The performance or efficiency of collections in Swift depends on several factors, including the type of collection, the collection size, the operations being performed on the collection, the way of implementation, and the compiler optimizations that are enabled. Higher performance implies that certain operations are faster and consume fewer resources, making them suitable for specific use cases. Lower performance indicates that some operations may be slower or resource-intensive in comparison. The table below shows the difference in performance of the most commonly used collections:

| Collection | Higher Performance | Lower Performance |
|------------|--------------------|-------------------|
| Array      | Random access: Arrays store their elements in contiguous memory, which makes it easy to access any element in the array in constant time. |  Inserting and deleting elements: this can require shifting the elements of the array in memory. |
| Dictionary |  Searching for elements: They store their key-value pairs using a hash table. | Random access: finding the key-value pair for a given key can take some time. |

> | Set        | Check for membership and find intersections, differences, and unions of sets: they store unique elements unordered. | Random access: elements are not stored in a contiguous memory location, so the entire set must be searched to access a specific element in a set. |

Reviewer: ```suggestion
| Set        | Checking for membership and finding intersections, differences, and unions of sets: they store unique elements unordered. | Random access: elements are not stored in a contiguous memory location, so the entire set must be searched to access a specific element in a set. |
```

## Benefits of Using Collections

Collections are fundamental data structures in Swift that provide various benefits, such as:

- **Organization and Management:** Collections provide a structured way to organize and manage multiple data items, making them easier to access, iterate, and manipulate.
- **Efficient Data Handling:** Collections are optimized for efficient data handling, making them suitable for storing and managing large amounts of data.
- **Declarative Syntax:** Swift's declarative syntax simplifies the creation and management of collections. This declarative approach reduces boilerplate code and improves code readability.
- **Algorithm Support:** Swift provides a rich collection of algorithms for manipulating collections, such as searching, sorting, and filtering.
- **Memory Management:** Collections are automatically managed by Swift's memory management system, ensuring efficient usage and preventing memory leaks.
- **Type Safety:** Swift's type system ensures that collections can only store elements of the same type. This prevents errors caused by incompatible data types and enhances code reliability.

> - **Interoperability:** Collections can be easily converted to other data structures, such as arrays or dictionaries, and vice versa. This creates a seamless work with data in different formats and allows integration with existing codebases.

Reviewer: ```suggestion
- **Interoperability:** Collections can be easily converted to other data structures, such as arrays or dictionaries, and vice versa. This can allow for seamless workflows when working with data in different formats and integration with existing codebases.
```
